### PR TITLE
New channels implementation for ORC

### PR DESCRIPTION
**Pull request by ringabout**
Thanks @mratsim !

The new implementation is mostly based on https://github.com/mratsim/weave/blob/5696d94e6358711e840f8c0b7c684fcc5cbd4472/unused/channels/channels_legacy.nim.

So I keep the license and the link.

Related article:
https://github.com/nim-lang/website/pull/274

### CODE REVIEW

File: lib/std/channel.nim [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+    Spsc: channel_close_spsc
+  ]
+
+  open_fn = [
+    Mpmc: channel_open_mpmc,
+    Mpsc: channel_open_mpsc,
+    Spsc: channel_open_spsc
+  ]
+
+proc channel_send(chan: ChannelRaw, data: sink pointer, size: int32, nonBlocking: bool): bool {.inline.} =
```
Comment (timotheecour) [inline lib/std/channel.nim]: style: `channelSend` (ditto for style with other symbols)

or is that intended?

File: lib/std/channel.nim [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+## 
+
+runnableExamples("-r:off --threads:on --gc:orc"):
+  # Be sure to compile with `--threads:on --gc:orc`.
+  import std/os
+
+  # In this example a channel is declared at module scope.
+  # Channels are generic, and they include support for passing objects between
+  # threads.
+  # Note that objects passed through channels will be deeply copied.
```
Comment (timotheecour) [inline lib/std/channel.nim]: also clarify whether ref objects are deeply copied or not

File: lib/std/channel.nim (hunk 1 above)
Comment (ringabout) [reply lib/std/channel.nim]: I'm not sure, but snake_case also appear in `std/atomics`

File: lib/std/channel.nim [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+    chan.send("Another message")
+
+  # Launch the worker.
+  var worker1: Thread[void]
+  createThread(worker1, firstWorker)
+
+  # Block until the message arrives, then print it out.
+  var dest = ""
+  chan.recv(dest) # "Hello World!"
+  echo dest
```
Comment (timotheecour) [inline lib/std/channel.nim]: turn into `assert dest == ...` (and then no need for the above comment `# "Hello World!"`)

File: lib/std/channel.nim [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+  #   Hello World!
+  #   Pretend I'm doing useful work...
+  #   Pretend I'm doing useful work...
+  #   Pretend I'm doing useful work...
+  #   Pretend I'm doing useful work...
+  #   Pretend I'm doing useful work...
+  #   Another message
+
+
+when not defined(gcArc) and not defined(gcOrc) and not defined(nimdoc):
```
Comment (timotheecour) [inline lib/std/channel.nim]: can we insead use: `when not defined(gcArc) and not defined(gcOrc):`
and use `nim doc --gc:arc lib/std/channels`
to avoid inconsistencies
eg:
```nim
when not defined(gcArc)
  proc foo*() = discard # this should be visible via `nim doc --gc:arc` but not by `nim doc`
```

File: lib/std/channel.nim (hunk 2 above)
Comment (ringabout) [reply lib/std/channel.nim]: I'm not sure

File: lib/std/channel.nim (hunk 4 above)
Comment (ringabout) [reply lib/std/channel.nim]: not sure whether it is worth.

File: lib/std/channel.nim (hunk 3 above)
Comment (ringabout) [reply lib/std/channel.nim]: fixed

File: lib/std/channel.nim [hunk 5]
```diff
@@ -0,0 +1,700 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2021 Mratsim & Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
```
Comment (timotheecour) [inline lib/std/channel.nim]: this needs a detailed doc comment or at least comment describing how this compares to lib/system/channels.nim, whether it'll evenually support --gc:refc, what to do with lib/system/channels.nim etc

File: lib/std/channel.nim [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+
+##
+## .. warning:: This module is experimental and its interface may change.
+##
+## The following is a simple example of two different ways to use channels:
+## blocking and non-blocking.
+## 
+
+runnableExamples("-r:off --threads:on --gc:orc"):
```
Comment (timotheecour) [inline lib/std/channel.nim]: can't this work without -r:off? what's preventing it?

File: lib/std/channel.nim [hunk 7]
```diff
@@ -0,0 +1,703 @@
...
+  var worker2: Thread[void]
+  createThread(worker2, secondWorker)
+  # This time, use a non-blocking approach with tryRecv.
+  # Since the main thread is not blocked, it could be used to perform other
+  # useful work while it waits for data to arrive on the channel.
+  while true:
+    var msg = ""
+    let tried = chan.tryRecv(msg)
+    if tried:
+      echo msg # "Another message"
```
Comment (timotheecour) [inline lib/std/channel.nim]: replace all echo's by something we can test, eg a global:
`var witness: int` (with a lock) or an atomic

File: lib/std/channel.nim (hunk 6 above)
Comment (ringabout) [reply lib/std/channel.nim]: it prints things, ok let me enable it.

File: lib/std/channel.nim (hunk 5 above)
Comment (ringabout) [reply lib/std/channel.nim]: It won't support refc. It will replace `lib/system/channels.nim` eventually when `orc` is production-ready(or when orc is default gc).

File: lib/std/channel.nim (hunk 7 above)
Comment (ringabout) [reply lib/std/channel.nim]: using echo make things more clear

Comment (timotheecour) [reply lib/std/channel.nim]: if using a global seq[string] (mimicking echo and collecting lines 1 by 1), you can test with `sort`

Comment (ringabout) [reply lib/std/channel.nim]: > The program should output something similar to this, but keep in mind that exact results may vary in the real world::

Comment (timotheecour) [reply lib/std/channel.nim]: but after `sort` we should have a deterministic output?

File: lib/std/channel.nim (hunk 1 above)
Comment (Araq) [reply lib/std/channel.nim]: Don't use `snake_case`.

File: lib/std/channel.nim (hunk 2 above)
Comment (Araq) [reply lib/std/channel.nim]: They are not copied, isolated subgraphs are *moved* around.

File: lib/std/channel.nim (hunk 7 above)
Comment (ringabout) [reply lib/std/channel.nim]: not really, the length of `messages` may vary.

Comment (ringabout) [reply lib/std/channel.nim]: I enable the test with non-strict assert.

File: lib/std/channel.nim (hunk 1 above)
Comment (timotheecour) [reply lib/std/channel.nim]: only `tcc_cas` and 6 ATOMIC_XXX symbols, all the rest is camlCase (including vast majority of stdib)

File: lib/std/channels.nim [hunk 8]
```diff
@@ -0,0 +1,697 @@
...
+
+##
+## .. warning:: This module is experimental and its interface may change.
+##
+## The following is a simple example of two different ways to use channels:
+## blocking and non-blocking.
+## 
+
+runnableExamples("--threads:on --gc:orc"):
+  # Be sure to compile with `--threads:on --gc:orc`.
```
Comment (timotheecour) [inline lib/std/channels.nim]: maybe remove, since redundant with `runnableExamples("--threads:on --gc:orc"):` (docs will show the command used `"--threads:on --gc:orc"`)

File: lib/std/channels.nim [hunk 9]
```diff
@@ -0,0 +1,697 @@
...
+  var worker2: Thread[void]
+  createThread(worker2, secondWorker)
+  # This time, use a non-blocking approach with tryRecv.
+  # Since the main thread is not blocked, it could be used to perform other
+  # useful work while it waits for data to arrive on the channel.
+  var messages: seq[string]
+  while true:
+    var msg = ""
+    let tried = chan.tryRecv(msg)
+    if tried:
```
Comment (timotheecour) [inline lib/std/channels.nim]:
```suggestion
    if chan.tryRecv(msg):
```

File: lib/std/channel.nim (hunk 1 above)
Comment (ringabout) [reply lib/std/channel.nim]: fixed

File: lib/std/channels.nim [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+    sleep(2000)
+    chan.send("Another message")
+
+  # Launch the worker.
+  var worker1: Thread[void]
+  createThread(worker1, firstWorker)
+
+  # Block until the message arrives, then print it out.
+  var dest = ""
+  chan.recv(dest)
```
Comment (timotheecour) [inline lib/std/channels.nim:53]: is there an API to block until either a message arrives or a timeout expires ?

Comment (ringabout) [reply lib/std/channels.nim:53]: not yet, they can be done in the future.

**Comment by ringabout** [comment]
This is the initial PR for new channel implementation. Docs and tests can be improved in the future.

**Comment by Araq** [comment]
Of course, but we need to decide if we need the other variants when "multi consumer, multi producer" can handle every case. Ping @mratsim. Do you think the *standard* library is better off with these distinctions? For comparison: We also try to not to offer too many different hash table types.

File: lib/std/channels.nim [hunk 11]
```diff
@@ -0,0 +1,703 @@
...
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  send_fn[chan.d.impl](chan.d, data.unsafeAddr, size, nonBlocking)
+
+proc channel_receive[T](chan: Chan[T], data: ptr T, size: int32, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recv_fn[chan.d.impl](chan.d, data, size, nonBlocking)
+
+func trySend*[T](c: Chan[T], src: sink Isolated[T]): bool {.inline.} =
```
Comment (Araq) [inline lib/std/channels.nim]: This is a problematic interface as when it returns `false` the message cannot be re-send. It should take a `var Isolated[T]` instead.

Comment (mratsim) [reply lib/std/channels.nim]: See https://github.com/nim-lang/RFCs/issues/347#issuecomment-791986889

File: lib/std/channels.nim [hunk 12]
```diff
@@ -0,0 +1,703 @@
...
+  ## Sends item to the channel(non blocking).
+  var data = src.extract
+  result = channel_send(c, data, int32 sizeof(data), true)
+  if result:
+    wasMoved(data)
```
Comment (Araq) [inline lib/std/channels.nim:658]: This shouldn't be required, `src.extract` has to empty the `src`. I think with `var Isolated` this problem disappears too.

File: lib/std/channels.nim [hunk 13]
```diff
@@ -0,0 +1,703 @@
...
+
+func tryRecv*[T](c: Chan[T], dst: var T): bool {.inline.} =
+  ## Receives item from the channel(non blocking).
+  channel_receive(c, dst.addr, int32 sizeof(dst), true)
+
+func send*[T](c: Chan[T], src: sink Isolated[T]) {.inline.} =
+  ## Sends item to the channel(blocking).
+  var data = src.extract
+  discard channel_send(c, data, int32 sizeof(data), false)
+  wasMoved(data)
```
Comment (Araq) [inline lib/std/channels.nim:485]: Shouldn't be required, `channel_send` should take a `sink` parameter. I think.

File: lib/std/channels.nim [hunk 14]
```diff
@@ -0,0 +1,697 @@
...
+#
+#           The Nim Compiler
+#        (c) Copyright 2021 Mratsim & Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+
+# Based on https://github.com/mratsim/weave/blob/5696d94e6358711e840f8c0b7c684fcc5cbd4472/unused/channels/channels_legacy.nim
```
Comment (mratsim) [inline lib/std/channels.nim:11]: Those are translations of @aprell (Andreas Prell) original channels from C to Nim (https://github.com/aprell/tasking-2.0/blob/master/src/channel_shm/channel.c)

And in turn they are an implementation of Michael & Scott lock-based queues (note the paper has 2 channels: lock-free and lock-based) with additional caching:
- Simple, Fast, and Practical Non-Blocking and Blocking Concurrent Queue Algorithms\
  Maged M. Michael, Michael L. Scott, 1996\
  https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf

File: lib/std/channels.nim [hunk 15]
```diff
@@ -0,0 +1,697 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2021 Mratsim & Nim Contributors
```
Comment (mratsim) [inline lib/std/channels.nim]:
```suggestion
#        (c) Copyright 2021 Andreas Prell, Mamy André-Ratsimbazafy & Nim Contributors
```

File: lib/std/channels.nim [hunk 16]
```diff
@@ -0,0 +1,697 @@
...
+    size: int32
+    itemsize: int32 # up to itemsize bytes can be exchanged over this channel
+    head: int32     # Items are taken from head and new items are inserted at tail
+    pad: array[CacheLineSize-sizeof(int32), byte] # Separate by at-least a cache line
+    tail: int32
+    buffer: ptr UncheckedArray[byte]
+
+  # TODO: Replace this cache by generic ObjectPools
+  #       We can use HList or a Table or thread-local globals
+  #       to keep the list of object pools
```
Comment (mratsim) [inline lib/std/channels.nim]: I don't think the TODO should be left here.

File: lib/std/channels.nim (hunk 13 above)
Comment (ringabout) [reply lib/std/channels.nim:485]: It did causes double free problems. I will try again,

File: lib/std/channels.nim [hunk 17]
```diff
@@ -0,0 +1,697 @@
...
+    raise newException(OutOfMemDefect, "Could not allocate memory")
+
+  initLock(result.headLock)
+  initLock(result.tailLock)
+  initCond(result.notFullCond)
+  initCond(result.notEmptyCond)
+
+  result.owner = -1 # TODO
+  result.impl = impl
+  result.closed.store(false, moRelaxed) # We don't need atomic here, how to?
```
Comment (mratsim) [inline lib/std/channels.nim:248]: The comment can be removed

File: lib/std/channels.nim [hunk 18]
```diff
@@ -0,0 +1,697 @@
...
+  send(c, isolate(src))
+
+func recv*[T](c: Chan[T], dst: var T) {.inline.} =
+  ## Receives item from the channel(blocking).
+  discard channelReceive(c, dst.addr, int32 sizeof(dst), false)
+
+func recvIso*[T](c: Chan[T]): Isolated[T] {.inline.} =
+  var dst: T
+  discard channelReceive(c, dst.addr, int32 sizeof(dst), false)
+  result = isolate(dst)
```
Comment (mratsim) [inline lib/std/channels.nim:498]: `recvIso` is strange, maybe we can just overload `recv`?

File: lib/std/channels.nim (hunk 11 above)
Comment (ringabout) [reply lib/std/channels.nim]: I see, thanks

File: lib/std/channels.nim [hunk 19]
```diff
@@ -0,0 +1,697 @@
...
+
+import std/[locks, atomics, isolation]
+import system/ansi_c
+
+# Channel (Shared memory channels)
+# ----------------------------------------------------------------------------------
+
+const
+  CacheLineSize {.intdefine.} = 64 # TODO: some Samsung phone have 128 cache-line
+  ChannelCacheSize* {.intdefine.} = 100
```
Comment (mratsim) [inline lib/std/channels.nim]: Do we want cachable channels?
They do work well in Weave and can handle trillions of Fibonacci tasks without issue so would work well as a primitive for Actor model inboxes or CSP channels.

But once allocated it's unclaimable memory.
At the very least it should be documented and probably renamed `NimChannelCacheSize`

File: lib/std/channels.nim [hunk 20]
```diff
@@ -0,0 +1,697 @@
...
+
+type
+  # ChannelBufKind = enum
+  #   Unbuffered # Unbuffered (blocking) channel
+  #   Buffered   # Buffered (non-blocking channel)
+
+  ChannelKind* = enum
+    Mpmc # Multiple producer, multiple consumer
+    Mpsc # Multiple producer, single consumer
+    Spsc # Single producer, single consumer
```
Comment (mratsim) [inline lib/std/channels.nim]: Maybe we only need MPMC channels for the stdlib and specialized channels should go in fusion or a library.

File: lib/std/channels.nim [hunk 21]
```diff
@@ -0,0 +1,697 @@
...
+    chanSize: int32
+    chanN: int32
+    chanKind: ChannelKind
+    numCached: int32
+    cache: array[ChannelCacheSize, ChannelRaw]
+
+# ----------------------------------------------------------------------------------
+
+template incmod(idx, size: int32): int32 =
+  (idx + 1) mod size
```
Comment (mratsim) [inline lib/std/channels.nim]: `mod` is a bit expensive, after that channel implementations I changed my base implementation of queues to avoid them without requiring power of 2 size, will do a recap.

File: lib/std/channels.nim (hunk 13 above)
Comment (ringabout) [reply lib/std/channels.nim:485]: I changed `channel_send` to `sink` parameter and removed `wasMoved`, the issue of yesterday appears.

**Comment by ringabout** [comment]
> but we need to decide if we need the other variants when "multi consumer, multi producer" can handle every case.

I can also unpublic related stuffs.

File: lib/std/channels.nim [hunk 22]
```diff
@@ -0,0 +1,697 @@
...
+    Mpmc # Multiple producer, multiple consumer
+    Mpsc # Multiple producer, single consumer
+    Spsc # Single producer, single consumer
+
+  ChannelRaw* = ptr ChannelObj
+  ChannelObj = object
+    headLock, tailLock: Lock
+    notFullCond: Cond
+    notEmptyCond: Cond
+    owner: int32
```
Comment (mratsim) [inline lib/std/channels.nim]: `owner` field is unused

File: lib/std/channels.nim [hunk 23]
```diff
@@ -0,0 +1,697 @@
...
+    impl: ChannelKind
+    closed: Atomic[bool]
+    size: int32
+    itemsize: int32 # up to itemsize bytes can be exchanged over this channel
+    head: int32     # Items are taken from head and new items are inserted at tail
+    pad: array[CacheLineSize-sizeof(int32), byte] # Separate by at-least a cache line
+    tail: int32
```
Comment (mratsim) [inline lib/std/channels.nim]: Can use `{.align: CacheLineSize.}` now

File: lib/std/channels.nim [hunk 24]
```diff
@@ -0,0 +1,697 @@
...
+
+  ChannelRaw* = ptr ChannelObj
+  ChannelObj = object
+    headLock, tailLock: Lock
+    notFullCond: Cond
+    notEmptyCond: Cond
+    owner: int32
+    impl: ChannelKind
+    closed: Atomic[bool]
+    size: int32
```
Comment (mratsim) [inline lib/std/channels.nim]: Do we expect more than 2 billions items enqueued in a channel?

**Review by mratsim — COMMENTED** [review]
Nitpicks:
- Some comments and TODOs shouldn't be here

API:
- `recvIso` is strange, `recv` ?
- We likely want to only expose the MPMC channels for the stdlib and specialized channel are for fusion or 3rd party libs
- Do we want the channel cache? it's a key ingredient for performance and handling very short tasks (that channel was used as a backend for Flowvars), but it's unreclaimable memory. At least it's size should be prefixed with "Nim" and it should be documented.
- Probably more for the RFC https://github.com/nim-lang/RFCs/issues/347, recently I have seen language having the following API for channel `let (producer, consumer) = newChan(size = 10)` so that the receiver and sender are using different types, do we want that.

Doc:
- Should mention the paper from Michael & Scott, since when perf is critical, people know the difference between the various concurrent queues and concurrency libraries propose many, see https://github.com/khizmax/libcds#references
  - People looking for advanced semantics like serializability and linearizability can use the paper as reference as well. 
- Should mention MPMC channels
- Should mention lock-based since progress guarantees are important

Internals:
- The channel is using `mod` in hot paths which is a very slow instruction
- The channel is allocated with "maxItems + 1" so that we have a sentinel slot to distinguish `isFull` and `isEmpty` CacheLineSize otherwise basic `(tail - head) mod size == 0` in both case.

This can be avoided as explained in https://github.com/mratsim/weave/blob/f41a562/weave/cross_thread_com/channels_spsc.md#litterature and https://github.com/elijahr/lockfreequeues/issues/1, and implemented in https://github.com/mratsim/weave/blob/f41a562/weave/cross_thread_com/channels_spsc.md#litterature.

Basically if we want to avoid mod:
- either we need a power of 2 size
- or we need an extra bit of information to distinguish the empty / full case
  - the extra bit can be provided by a sentinel object, potentially wasting a lot of space if the object is big and queue is expected small (say bounded to 4 items)
  - the extra bit can be a bool but it now becomes a new contention point that is both read and written
  - We make the counters use 2x more size!

File: lib/std/channels.nim (hunk 16 above)
Comment (ringabout) [reply lib/std/channels.nim]: fixed

File: lib/std/channels.nim (hunk 22 above)
Comment (ringabout) [reply lib/std/channels.nim]: fixed

File: lib/std/channels.nim (hunk 17 above)
Comment (ringabout) [reply lib/std/channels.nim:248]: removed unused `owner`.

File: lib/std/channels.nim (hunk 14 above)
Comment (ringabout) [reply lib/std/channels.nim:11]: added

File: lib/std/channels.nim (hunk 24 above)
Comment (Araq) [reply lib/std/channels.nim]: hardly but `space: int` would be more idiomatic.

File: lib/std/channels.nim (hunk 18 above)
Comment (ringabout) [reply lib/std/channels.nim:498]: I'm not sure whether `recv(c: Chan[T]): T` is needed.

File: lib/std/channels.nim (hunk 23 above)
Comment (ringabout) [reply lib/std/channels.nim]: Do you mean this?
`head {.align: CacheLineSize.} : int`

File: lib/std/channels.nim (hunk 24 above)
Comment (ringabout) [reply lib/std/channels.nim]: I use `int`.

File: lib/std/channels.nim (hunk 20 above)
Comment (ringabout) [reply lib/std/channels.nim]: I unpublic all kind related procs and they may be reused in the future.

File: lib/std/channels.nim (hunk 19 above)
Comment (ringabout) [reply lib/std/channels.nim]: renamed

File: lib/std/channels.nim (hunk 23 above)
Comment (mratsim) [reply lib/std/channels.nim]: The tail should to be on a different cache line from the head (and each with their lock on the same cache line).

We could also make the buffer and size on yet another cacheline for common data.

File: lib/std/channels.nim (hunk 21 above)
Comment (ringabout) [reply lib/std/channels.nim]: can be improved in the future.

File: lib/std/channels.nim (hunk 23 above)
Comment (ringabout) [reply lib/std/channels.nim]: I see, thanks

**Review by dom96 — COMMENTED** [review]
This looks to have an API similar to our existing channels, do we really need a separate module for this?

If we do then we should probably name it something other than `channels` since that is what the existing module is called already.

**Comment by Araq** [comment]
> If we do then we should probably name it something other than channels since that is what the existing module is called already.

The existing one isn't an import...

File: tests/stdlib/tchannel_pthread.nim [hunk 25]
```diff
@@ -0,0 +1,334 @@
+discard """
```
Comment (timotheecour) [inline tests/stdlib/tchannel_pthread.nim]: rename tests/stdlib/tchannel_pthread.nim => tests/stdlib/tchannels_pthread.nim

ditto with other test file

File: lib/std/channels.nim [hunk 26]
```diff
@@ -0,0 +1,693 @@
+#
```
Comment (timotheecour) [inline lib/std/channels.nim:1]: can we rename
```
lib/system/channels.nim
=>
lib/system/channels_builtin.nim
```
to avoid human confusion? (even though it's an include)
EDIT: also to avoid conflict in docs, since as noted by @dom96 we already generate channels.html so there would be a conflict

**Comment by timotheecour** [comment]
is there anything preventing std/channels from being usable with gc:refc (not necessarily now but in future work) ?
at least a subset of the API or for a subset of the types being sent/received

if so then API's in system/channels could be eventually deprecated and removed

**Comment by dom96** [comment]
>> If we do then we should probably name it something other than channels since that is what the existing module is called already.

> The existing one isn't an import...

Yes, but it gets its own documentation page titled "channels": https://nim-lang.org/docs/channels.html

I don't see how you'll get around renaming it to be honest.

**Comment by timotheecour** [comment]
> I don't see how you'll get around renaming it to be honest.

as suggested in https://github.com/nim-lang/Nim/pull/17305#discussion_r591891512, we should rename the old one to `lib/system/channels_builtin.nim` (we can bikeshed the name), so that there's no conflict in the docs; note that it's generated from kochdocs as a special case because it's an include:
```nim
  doc0 = """
lib/system/threads.nim
lib/system/channels.nim
""".splitWhitespace() # ran by `nim doc0` instead of `nim doc`
```
(in future work, we should if possible turn `lib/system/channels_builtin.nim` into an import, possibly a private import in lib/std/private/channels_builtin.nim`)

**Comment by dom96** [comment]
Sure, we can do that, but it's not ideal as existing URLs to that doc will break. Oh well I guess, I sense that there is no appetite to rename this module (even though it is new...)

**Comment by timotheecour** [comment]
> existing URLs to that doc will break

only these need updates within nim repo.

```
doc/lib.rst:42:14:* `channels <channels.html>`_
doc/manual.rst:7723:50:See the `threads <threads.html>`_ and `channels <channels.html>`_ modules
lib/pure/collections/deques.nim:49:24:## * `channels module <channels.html>`_ for inter-thread communication
lib/pure/concurrency/threadpool.nim:17:24:## * `channels module <channels.html>`_ for message passing support
```

urls eternal to nim repo pointing to channels.html will break, but I don't think we should offer that kind of stability (prevents refactorings such as importing / re-exporting symbols, etc)

**Comment by ringabout** [comment]
see https://github.com/nim-lang/Nim/pull/17330

File: tests/stdlib/tchannel_pthread.nim (hunk 25 above)
Comment (ringabout) [reply tests/stdlib/tchannel_pthread.nim]: fixed

File: lib/std/channels.nim [hunk 27]
```diff
@@ -0,0 +1,693 @@
...
+
+func open*[T](c: Chan[T]): bool {.inline.} =
+  result = c.d.channel_open()
+
+func close*[T](c: Chan[T]): bool {.inline.} =
+  result = c.d.channel_close()
+
+func peek*[T](c: Chan[T]): int {.inline.} = peek(c.d)
+
+proc initChan*[T](elements = 30): Chan[T] =
```
Comment (timotheecour) [inline lib/std/channels.nim]: * IMO we shouldn't abbreviate it, `Channel[T]` is short enough (it's not like it's gonna appear thousands of times in your code)
* ditto, `initChan` => `initChannel`

Comment (ringabout) [reply lib/std/channels.nim]: make sense

Comment (ringabout) [reply lib/std/channels.nim]: fixed

File: lib/std/channels.nim [hunk 28]
```diff
@@ -0,0 +1,511 @@
...
+
+func open*[T](c: Channel[T]): bool {.inline.} =
+  result = c.d.channelOpenMpmc()
+
+func close*[T](c: Channel[T]): bool {.inline.} =
+  result = c.d.channelCloseMpmc()
+
+func peek*[T](c: Channel[T]): int {.inline.} = peek(c.d)
+
+proc newChannel*[T](elements = 30): Channel[T] =
```
Comment (timotheecour) [inline lib/std/channels.nim:508]: * document that `elements = 1` will cause channel to be unbuffered
* magic number 30 sounds weird; how about making the default unbuffered instead (and using an impossible value, say -1, to denote that the channel is unbufferred)
* isn't typedesc param more common now-days? [T] is mostly useful for implicit generic instantiation

Comment (ringabout) [reply lib/std/channels.nim:508]: make sense

**Comment by timotheecour** [comment]
/cc @mratsim 
why are we using something like a `BoundedQueue` (refs https://github.com/mratsim/weave/blob/2e528bd2a6a04306fd029f04a0af87fa201e5d89/weave/datatypes/bounded_queues.nim#L42) instead of a std/deques?

with current API, you'd need to know ahead of time the capacity to use, which is in practice really hard: we have an impossible dilemma to solve:
* either use a capacity that's too large (over-provisions, wastes memory)
* or use a capacity that's too small (under-provisions, causes wait's when buffer fills up, affecting performance)

with std/queues, you avoid this impossible dilemma and let the buffer grow as needed. We could improve this by adding a maximum cap on the deques underlying buffer but that's a minor addition to an existing module.

Furthermore, even if somehow a `BoundedQueue` would be better, IMO that part could be factored out and reused, it'd be useful not just in the context of std/channels

**Review by timotheecour — APPROVED** [review]
LGTM to make progress, with no expectation of API stability at this point.

I'm really curious about whether we should instead use std/dequeues, and a few other points, but can be addressed in followup work

**Comment by Araq** [comment]
> with std/queues, you avoid this impossible dilemma and let the buffer grow as needed. We could improve this by adding a maximum cap on the deques underlying buffer but that's a minor addition to an existing module.

In my experience this is the far better design than unbounded queues, these tend to consume way too much memory and you lose the flow control aspect of bounded queues. Golang also uses bounded queues as far as I know and never revisited this design decision.

**Comment by mratsim** [comment]
> /cc @mratsim
> why are we using something like a `BoundedQueue` (refs https://github.com/mratsim/weave/blob/2e528bd2a6a04306fd029f04a0af87fa201e5d89/weave/datatypes/bounded_queues.nim#L42) instead of a std/deques?
> 
> with current API, you'd need to know ahead of time the capacity to use, which is in practice really hard: we have an impossible dilemma to solve:
> 
>     * either use a capacity that's too large (over-provisions, wastes memory)
> 
>     * or use a capacity that's too small (under-provisions, causes wait's when buffer fills up, affecting performance)
> 
> 
> with std/queues, you avoid this impossible dilemma and let the buffer grow as needed. We could improve this by adding a maximum cap on the deques underlying buffer but that's a minor addition to an existing module.
> 
> Furthermore, even if somehow a `BoundedQueue` would be better, IMO that part could be factored out and reused, it'd be useful not just in the context of std/channels

There are multiple reasons:

1. Very often, either the producer is much faster than the consumer or the consumer is much faster than the producer.\
    There are 3 ways to handle the first case at the channel level, buffering, blocking or dropping.
    - Buffering is the unbounded queue solution you propose, remember that ultimately they are bounded by your memory
    - Dropping would be a ring buffer with the oldest message being overwritten
    - Blocking is the bounded queue
    
    Additionally, the producer can throttle itself and adapt it's rate of message, which assumes it has some feedback about the consumer consumption speed.
    
    I think we can all agree that dropping is too niche for a standard library solution which leaves discussing buffering or blocking.
    
    Buffering is an out-of-memory waiting to happen, with no recourse for the producer to detect if it's too fast. People **need** to think about their buffer size in a multithreading scenario and bounded queues are needed for backpressure management. This allows producers to adapt and potentially throttle themselves in case the consumer is overloaded. Usually, the messages are coming from IO (say networking), and passed via channels to a compute thread, blocking channels are the way for compute thread to communicate that it's overloaded and that messages cannot be handled so an alternative should be found (maybe throttle the source, maybe deactivate non-essential services/computations, maybe ban it because it's a DOS, ...).

    On the other hand, you can easily emulate an unbounded channel from a bounded channel by using `trySend` and if it blocks, putting the data in your own buffer in front of the channel. It's also quite possible that the buffering shouldn't happen in the channel or just in front of it but more upstream in the "offending" producer because multiple-producer can enqueue messages in the channel and only one can misbehave, or some should have higher priorities than other in case of congestion.

2. You can't use the 2 locks solution with unbounded channels because on reaching the bound, the producer must lock the consumer out while resizing. Hence you get higher lock contention and worse performance.
3. When you need unbounded channels, usually you use list-based channels.

References on backpressure:
- https://medium.com/@jayphelps/backpressure-explained-the-flow-of-data-through-software-2350b3e77ce7
- https://lucumr.pocoo.org/2020/1/1/async-pressure/

**Comment by timotheecour** [comment]
you're comparing `BoundedQueue` vs unbounded `std/dequeues`, but I was referring to `std/dequeues` with a max capacity beyond which you'd fallback to essentially the same as a `BoundedQueue`.

* the initial size parameter becomes irrelevant, and doesn't even need to be a user parameter, could be set to a fixed small value (say 0) as initial queue capacity
* the capacity of the dequeue keeps doubling (as in deques) as needed until a max capacity is reached
* then it falls back to the `BoundedQueue` behavior

This gives best of both worlds:
* no risk of over or under provision
* no unbounded memory use, in fact same memory use as `BoundedQueue` in the worst case; the max capacity parameter (the only parameter) is not very sensitive since if you overprovision it, you won't pay the memory cost unless you actually need it (eg: producer faster than consumer)
* as future refinement, it could also shrink using standard algorithm (2 hysteresis thresholds) which can be useful to dynamically adapt while maintaining resource usage proportional to current needs in a long running scenario

**Comment by Araq** [comment]
We don't have an implementation for your "best of both worlds" though and it's not clear how it would perform.

**Comment by mratsim** [comment]
`std/dequeues` does over-provision because it requires power of 2 size, so if you use 33 as your capacity, you will allocate for 64 slots and pay for that cost. The current channels proposed allocate exactly what the user wants and provide excellent performance and overhead characteristics.

Now `std/dequeues` implementation can be improved to use the current PR backend implementation but that's for a separate PR I think.

For specialized needs, the producer can compose the base channel with a buffer in front that provides the extra leeway needed, they can choose a small channel and implement a buffer with the hysteresis strategy you mention as ans example. 

In general, all the dynamism and complex logic should be done outside of the base channel.
Alternatively they might need to write their own channels, but it's easier to swap a simple channel and reuse their additional logic than reimplement all of what they did on some channels.

I've mentioned it in the channels API RFC PR https://github.com/nim-lang/RFCs/issues/347#issue-822547045
There is no "2 worlds" regarding channels, we should provide a good default and compose well because it is impossible to write a channel that satisfies all the potential expectations of a channel:
- single-producer or multi-producer
- single-consumer or multi-consumer
- list-based or array-based or hybrid
  - If list-based:
    - intrusive linked list or non-intrusive
- bounded or unbounded
  - if bounded:
    - Errors on reaching bounds
    - Blocks on reaching bounds
    - Overwrite the oldest item on reaching bounds (ring-buffer)
- lock-based or atomic-based
- progress guarantees:
  - no guarantees (whether using locks or atomics)
  - obstruction-free
  - lockfree
  - waitfree
- priorities support
- item ordering guarantees
  - no guarantees
  - serializability
  - linearizability
- Buffered or unbuffered (rendez-vous)

**Comment by Araq** [comment]
> is there anything preventing std/channels from being usable with gc:refc (not necessarily now but in future work) ?
at least a subset of the API or for a subset of the types being sent/received

I think there is no good way to make these work with the old GCs. We will eventually make ``--gc:orc`` the default for a reason. ;-)

> if so then API's in system/channels could be eventually deprecated and removed

That's true.

**Comment by timotheecour** [comment]
> std/dequeues does over-provision because it requires power of 2 size, so if you use 33 as your capacity, you will allocate for 64 slots and pay for that cost

it doesn't over-provision, you do as follows:
given user provided max capacity C (C = 33 in your case):
you double capacity as needed (starting from 1) until it reaches the max power of 2 <= C (ie, 32 in your case), and when this fills up, you increase the length 1 final time to C ( = 33) and switch to `BoundedQueue` regime.

so yes, you do get best of both worlds: no overprovisioning and pay-what-you-actually-use, instead of a max capacity.

performance wise, it's at least as good as `BoundedQueue`.

**Comment by Araq** [comment]
Merging for now, we can refine it later.
